lib/msf/core/exploit/remote/ldap.rb

```diff
@@ -64,22 +98,46 @@ def get_connect_opts
       connect_opts
     end
 
+    #
+    # Connect to the target LDAP server using the options provided,
+    #   and pass the resulting connection object to the proc provided.
+    #   Terminate the connection once the proc finishes executing.
+    #
+    # @param opts [Hash] Options for the LDAP connection.
+    # @param &block [Proc] A proc containing the functionality to execute
+    #   after the LDAP connection has succeeded. The connection is closed
+    #   once this proc finishes executing.
+    # @see Net::LDAP.open
+    # @return [Nil] Nothing is returned by this function.
```

Comment:
Doesn't this return an `Net::LDAP` instance?

Reply:
Nope it doesn't. It just passes a Net::LDAP instance to the block that is passed in. The function itself doesn't return anything to the best of my knowledge. See https://www.rubydoc.info/gems/ruby-net-ldap/Net%2FLDAP:open for more info. Note that doesn't say that it returns any data but rather that it just yields to the block provided.

Reply:
It looks like this method implicitly returns the result of the block passed in - and not `nil` 👀

```
result = ldap_connect do |ldap|
  ldap
end

puts result
# => #<Net::LDAP:0x00007faeb9b73490 ...>
```

Reply:
Huh thanks for pointing this out @adfoster-r7 looks like I made an oversight here. I'll have to see how to reflect this properly within YARD.

Reply:
I think you're leaking implementation details by swapping this over to an array of hashes, i.e. a module might pass in a block that returns something arbitrary.

It'd probably be simpler to go with something like:

```
# @yieldparam [Net::LDAP] ldap The current Net::LDAP instance
# @return [Object] The result of the yielded block
```

Or, define the API yourself instead and always return nil

```
# ... yield docs here ...
# @return [Nil] Nothing is returned by this function.
def ldap_connect(opts = {}, &block)
  Net::LDAP.open(get_connect_opts.merge(opts), &block)
  nil
end
```

Note: I just eyeballed the syntax from the yard docs site - probably worth verifying the output is as expected - https://rubydoc.info/gems/yard/file/docs/Tags.md#yieldparam

Reply:
> I think you're leaking implementation details by swapping this over to an array of hashes, i.e. a module might pass in a block that returns something arbitrary.
> 
> It'd probably be simpler to go with something like:
> 
> ```ruby
> # @yieldparam [Net::LDAP] ldap The current Net::LDAP instance
> # @return [Object] The result of the yielded block
> ```
> 

This looks good to me and I'll go ahead and adjust the function to this as I think that is more accurate.

As for the `@yeildparam` did that comment just apply to this function or to other sections as well?

Reply:
Note for future travelers whilst I was looking at this, but `@yeildparam` is meant to be on a function that is expected to be passed a block and then yield to that block.

For instance here we have a function that is declared with `@yieldparam` to return `Net::LDAP` object:

![image](https://user-images.githubusercontent.com/63261883/217962105-862651d6-8845-4546-9415-1a6d4fa1720f.png)

As one can see this is denoted in RubyMine as a undeclared `&blk` parameter that is expected to be passed in and to which we set its type to `Proc<Net::LDAP>`, aka a proc that is expected to return a Net::LDAP object.

Hope that helps explain when to use it a bit clearer. Found this whilst reading https://blog.jetbrains.com/ruby/2019/06/yard-support-in-rubymine/#yield_parameters and decided to do some testing to confirm this definition didn't need updating; looks like it was used correctly here.